pkg/reconciler/build/timeouts_handler.go

```diff
@@ -0,0 +1,81 @@
+package build
+
+import (
+	"time"
+
+	clientset "github.com/knative/build/pkg/client/clientset/versioned"
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const checkInterval = 10 * time.Second
+
+// TimeoutSet contains required objects to handle build timeouts
+type TimeoutSet struct {
+	logger         *zap.SugaredLogger
+	kubeclientset  kubernetes.Interface
+	buildclientset clientset.Interface
+	stopCh         <-chan struct{}
+}
+
+// NewTimeoutHandler returns TimeoutSet filled structure
+func NewTimeoutHandler(logger *zap.SugaredLogger,
+	kubeclientset kubernetes.Interface,
+	buildclientset clientset.Interface,
+	stopCh <-chan struct{}) *TimeoutSet {
+	return &TimeoutSet{
+		logger:         logger,
+		kubeclientset:  kubeclientset,
+		buildclientset: buildclientset,
+		stopCh:         stopCh,
+	}
+}
+
+// TimeoutHandler walks through all namespaces, gets list of builds and checks timeouts
+func (t *TimeoutSet) TimeoutHandler() error {
+	ticker := time.NewTicker(checkInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-t.stopCh:
+			return nil
+		case <-ticker.C:
```

Comment:
Rather than polling for timeouts, what do you think about having a goroutine per ongoing build that does:

```go
select {
  case <-t.stopCh:
    // controller stopping, give up
  case <- build.finished:
    // build finished, give up
  case <- time.After(timeoutDur):
    // timeout elapsed
    t.stopBuild(build)
}
```

This should stop builds ~immediately upon timing out, which is nice, but also avoids unnecessary polling when we can tell the timeout is still minutes (or hours!) away.

We'd also need to spawn these goroutines when the controller starts too, in case the controller dies and restarts, to make sure ongoing builds have their timeouts enforced.